### DISCUSSION TITLE
Connect local ws server to browser frontend

### DISCUSSION THREAD

**Discussion by brianignacio5** [eclipse-theia/theia#9252] [Q&A] [answered]
Hi there! Anyone can help me to understand Theia a bit better?

Let's say we have a browser app version of Theia and the user has a web socket server running locally. How would you connect the frontend of a Theia extension to this web socket server ? Maybe point me to an example application ? I tried using the json-rpc doc in https://theia-ide.org/docs/json_rpc but not sure how to test it is working.

**Comment by paul-marechal** [comment, marked as answer]
If your code runs in the browser you can simply use the browser's WebSocket APIs and whatever protocol your server uses. If it is running in node you can depend on the `ws` package to do the same but using a different API.

**Comment by brianignacio5** [comment]
Thanks @marechal-p I analyzed the backend template from theia generator to understand how it works.

**Reply by paul-marechal** [reply]
See https://eclipsesource.com/blogs/2019/11/21/how-to-create-develop-an-eclipse-theia-ide-extension/ for help developing Theia extensions. Note that you can write code to run in the browser and/or the Node.js server, depending on what you need.
